### ISSUE TITLE
Incompatible with VSCode 1.5.0+

### ISSUE THREAD

**Issue by renehamburger** [jcward/vscode-hxcpp-debug#44]
I haven't managed to get this extension to run on VSCode 1.8.1 on Windows 10. Is your [example launch.json](https://github.com/jcward/vscode-hxcpp-debug/blob/master/test%20openfl/.vscode/launch.json) still up-to-date? The compiler complains about the "cwd" property being missing, for example.
It also complains about the launch type "hxcpp" being unknown, even though the extension seems to have been installed correctly. The built of the extension passed and it's displayed as installed within VSCode. Is it possible that [your package.json](https://github.com/jcward/vscode-hxcpp-debug/blob/master/package.json) needs updating, too?

**Comment by Jarrio** [comment]
Personally I've tried messing with the package.json and couldn't get any difference

**Comment by jcward** [comment]
The `/tmp/adator.log` file shows:

```
1483975710.46462: Reading 241 bytes...
1483975710.4647: {"command":"initialize","arguments":{"adapterID":"hxcpp","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true},"type":"request","seq":1}
1483975710.46473: Got command: initialize
1483975710.46474: Initializing...
1483975710.46475: Sending 55 bytes:
1483975710.46475: {"command":"initialize","success":true,"request_seq":1}
```

Maybe they changed the handshake? Seems like an odd thing to do...

**Comment by hoseyjoe** [comment]
I never have been able to get hxcpp debugger working but I did notice in 1.8.1 removing 
`

            "runtimeExecutable": "${execPath}",
			"args": [
                "compilePath=${workspaceRoot}",
                "compileCommand=nme build windows project.nmml",
                "runPath=${workspaceRoot}/build/nme/windows/project/",
                "runCommand=project",
                "runInTerminal=false"],		`	

Makes "type":"hxcpp" acceptable...IDK

**Comment by jcward** [comment]
Just leaving myself some comments:

- The initialization sequence (or communication protocol) changed somehow. vscode does not respond to my response to the "initialize" command.
- I noticed a new "initialized" event, but it gets sent after the "launch" command, and we're not getting there yet.
- Looking at .ts source / examples: [debugProtocol.ts](https://github.com/Microsoft/vscode-debugadapter-node/blob/master/protocol/src/debugProtocol.ts), [debugSession.ts](https://github.com/Microsoft/vscode-debugadapter-node/blob/master/adapter/src/debugSession.ts), and [goDebug.ts](https://github.com/Microsoft/vscode-go/blob/master/src/debugAdapter/goDebug.ts)
- The [limited docs](https://code.visualstudio.com/docs/extensionAPI/api-debugging#_the-vs-code-debug-protocol-in-a-nutshell) on the debug protocol state:
> If your adapter is derived from the TypeScript or C# default implementation DebugSession, you don't have to handle the initialize request yourself.
- Great, so that means they can change it without much notice.
- The break occurred between 1.4 (working) and 1.5.0 (no longer working), both available from the [updates page](https://code.visualstudio.com/updates/July_2016). Note that 1.5 is packaged for linux differently, in a .tar.gz file instead of a .zip. TODO: investigate changes between 1.4 and 1.5.0 more closely.

**Comment by weinand** [comment]
Since we are still on version 1.x of the debug protocol we are not allowed to break the protocol (and we have not heard of any broken debugger since we've shipped VS Code 1.0). However, we have added quite a few features to the debug protocol in a backward compatible way.

In order to verify that nothings breaks when we change or add something to the debug protocol, I use the mono-debug extension which does not rely on the JS/TS debugadapter npm module:  https://github.com/Microsoft/vscode-mono-debug

I've reviewed the change history but I could not find any commit that would explain the problem you are seeing.

I tried to build your haxe debug extension, but the build steps are incomplete (I do not have haxe installed).

I quickly looked into your implementation but I could not find that your debug adapter sends a "InitializedEvent" to VS Code. It could be that earlier versions of VS Code were more resilient if this event was missing. But it was always a required event. Here is the corresponding doc:

> Since VS Code persists breakpoints on behalf of the debug adapter, it has to register the breakpoints with the debug adapter when a session starts. Since VS Code does not know when is a good time for this, the debug adapter is expected to send an initialize event to VS Code to announce that it is ready to accept breakpoint configuration requests.

I suggest to send this event at the end of your "initialize" handler (after you have sent the "initialize" response). See how this is done in mono-debug or mock-debug.

**Comment by jcward** [comment]
@weinand -- thank you so much for the detailed response. I saw InitializedEvent -- but in the Go adapter, it appears to be [sending InitializedEvent](https://github.com/Microsoft/vscode-go/blob/master/src/debugAdapter/goDebug.ts#L410) in the launchRequest handler, not in the [initializeRequest handler](https://github.com/Microsoft/vscode-go/blob/master/src/debugAdapter/goDebug.ts#L353)... Is this not correct? In the mock, I saw it in the initializeRequest handler, but I assumed this was because mock doesn't actually initialize anything. And the Go schema made more sense -- until my app is launched, I can't actually handle breakpoints (though I could queue them, maybe.)

I'm pretty sure I also tried sending an InitializedEvent in my InitializeRequest handler to no effect, but I'll double check. I did also notice that if I set a breakpoint in VSCode, a different sequence of init events happens and leads to a bad state. I'll try to characterize that further as well.

Thanks again for your time.

**Comment by weinand** [comment]
@jcward The "InitializedEvent" triggers a whole sequence of requests in VS Code, one being the request to set breakpoints in your debug adapter. It depends on your debug adapter when this should happen. The doc says: "Since VS Code does not know when is a good time for this, the debug adapter is expected to send an initialize event to VS Code to announce that it is **ready to accept breakpoint configuration requests.**" 
node-debug has the same problem as hack-debug: it cannot accept breakpoints before the node programm is launched. So node-debug sends the "InitializedEvent" late.

**Comment by jcward** [comment]
@weinand - I've created a [simple Ruby script](https://hastebin.com/qejeyekalo.rb) that mimics the init behavior of my debug adapter to test initialization sequences. I'm working on Linux, but it should also work on OSX. You can place it in my debug adapter at `vscode-hxcpp-debug/bin/linux/vschxcppdb` (and ensure it's executable, `chmod a+x vschxcppdb`)

It logs all incoming commands at `/tmp/echo.log` and responds to the initialize command with a success message. You can comment or uncomment the `send_event({"event"=>"initialized"})` line.

There are 4 logs below -- in vscode 1.4 and 1.8, each with and without the initialized events.

In VSCode 1.4, with or without the initialized event, vscode sends a "launch" command right after the InitializeResponse.

In VSCode 1.8 (and anything after 1.5), the behavior is: if I don't send an initialized event, it hangs and does not respond. If I do send an initialized event, it requests my threads. **What?!** How is it legitimate for VSCode to request threads when it hasn't sent a launch command yet?

So if I don't send an initialized event, it hangs, and if I do, it asks for thread before launching. What am I to do?

<hr>

VSCode 1.4, without the initialized event:
```
Got message: {"type"=>"request", "seq"=>1, "command"=>"initialize", "arguments"=>{"adapterID"=>"hxcpp", "pathFormat"=>"path", "linesStartAt1"=>true, "columnsStartAt1"=>true, "supportsVariableType"=>true, "supportsVariablePaging"=>true}}
Sending 32 bytes: {"success":true,"request_seq":1}
Got message: {"type"=>"request", "seq"=>2, "command"=>"launch", "arguments"=>{"name"=>"OSX/Linux: Debug", "args"=>["runPath=/home/jward/dev/vscode-hxcpp-debug/test cli/export", "runCommand=Main-debug", "runInTerminal=true"], "type"=>"hxcpp", "request"=>"launch", "program"=>"", "stopOnEntry"=>true}}
WARNING: Got unknown command: launch, ignoring...
```

VSCode 1.4, with the initialized event:
```
Got message: {"type"=>"request", "seq"=>1, "command"=>"initialize", "arguments"=>{"adapterID"=>"hxcpp", "pathFormat"=>"path", "linesStartAt1"=>true, "columnsStartAt1"=>true, "supportsVariableType"=>true, "supportsVariablePaging"=>true}}
Sending 32 bytes: {"success":true,"request_seq":1}
Sending 46 bytes: {"event":"initialized","seq":1,"type":"event"}
Got message: {"type"=>"request", "seq"=>2, "command"=>"launch", "arguments"=>{"name"=>"OSX/Linux: Debug", "args"=>["runPath=/home/jward/dev/vscode-hxcpp-debug/test cli/export", "runCommand=Main-debug", "runInTerminal=true"], "type"=>"hxcpp", "request"=>"launch", "program"=>"", "stopOnEntry"=>true}}
WARNING: Got unknown command: launch, ignoring...
```

VSCode 1.8 without the initialized event:
```
Got message: {"command"=>"initialize", "arguments"=>{"adapterID"=>"hxcpp", "pathFormat"=>"path", "linesStartAt1"=>true, "columnsStartAt1"=>true, "supportsVariableType"=>true, "supportsVariablePaging"=>true, "supportsRunInTerminalRequest"=>true}, "type"=>"request", "seq"=>1}
Sending 32 bytes: {"success":true,"request_seq":1}
```

VSCode 1.8 with the initialized event:
```
Got message: {"command"=>"initialize", "arguments"=>{"adapterID"=>"hxcpp", "pathFormat"=>"path", "linesStartAt1"=>true, "columnsStartAt1"=>true, "supportsVariableType"=>true, "supportsVariablePaging"=>true, "supportsRunInTerminalRequest"=>true}, "type"=>"request", "seq"=>1}
Sending 32 bytes: {"success":true,"request_seq":1}
Sending 46 bytes: {"event":"initialized","seq":1,"type":"event"}
Got message: {"command"=>"threads", "type"=>"request", "seq"=>2}
WARNING: Got unknown command: threads, ignoring...
```

**Comment by weinand** [comment]
With mock-debug I see this sequence of requests and events:
```
got initializeRequest
send InitializedEvent
got launchRequest
got threadsRequest
got threadsRequest
got stackTraceRequest
got scopesRequest
got scopesRequest
got scopesRequest
got scopesRequest
```

With the exception of the two threadsRequest this looks correct to me.

**Comment by jcward** [comment]
Ok, luckily debugging release versions of VSCode is super easy with command `Developer: Toggle Developer Tools`. I found my first problem. In VSCode rev `68ac40`, message.type [became required](https://github.com/Microsoft/vscode/commit/68ac04b#diff-0e4ee0da1500c71de4cddbaa995f47e3R157) (before it was implied as `response` if not an `event`.) This is my fault, as the mock debug message type has always said `type` is a required field. Easy enough, fixed in a07ad8c.

But I'm not closing this yet -- there is still a problem with Thread ID's, I'm getting `ErrorNoSuchThread(0)`. Perhaps 0 is an invalid thread ID? Investigating...

**Comment by weinand** [comment]
@jcward great to hear that you could resolve the issues!
